File: problem-solving/adventofcode2020/day11.py
# https://adventofcode.com/2020/day/11
with open('./day11.txt', 'r') as f:
    seats = [list(line) for line in f.read().splitlines()]


def get_adjacent_seats_count(seats, row, col):
    count = 0
    eight_directions = [(-1, -1), (-1, 0), (0, -1), (1, 1),
                        (1, 0), (0, 1), (-1, 1), (1, -1)]
    for d in eight_directions:
        diff_row, diff_col = d[0], d[1]
        r = row+diff_row
        c = col+diff_col
        if r >= 0 and r < len(seats) and c >= 0 and c < len(seats[0]):
            seat = seats[r][c]
            if seat == '#':
                count += 1
    return count


def get_eight_direction_seats_count(seats, row, col):
    count = 0
    eight_directions = [(-1, -1), (-1, 0), (0, -1), (1, 1),
                        (1, 0), (0, 1), (-1, 1), (1, -1)]
    for d in eight_directions:
        diff_row, diff_col = d[0], d[1]
        r = row+diff_row
        c = col+diff_col
        while r >= 0 and r < len(seats) and c >= 0 and c < len(seats[0]):
            seat = seats[r][c]
            if seat == '#':
                count += 1
                break
            elif seat == 'L':
                break
            else:
                r += diff_row
                c += diff_col

    return count


def get_seats_after_one_round(seats):
    new_seats = []
    for r in range(0, len(seats)):
        new_row = []
        for c in range(0, len(seats[r])):
            seat = seats[r][c]
            if seat == "L":
                if get_adjacent_seats_count(seats, r, c) == 0:
                    new_row.append("#")
                else:
                    new_row.append("L")
            elif seat == "#":
                if get_adjacent_seats_count(seats, r, c) >= 4:
                    new_row.append("L")
                else:
                    new_row.append("#")
            else:
                new_row.append(".")
        new_seats.append(new_row)
    return new_seats


def get_seats_after_one_round2(seats):
    new_seats = []
    for r in range(0, len(seats)):
        new_row = []
        for c in range(0, len(seats[r])):
            seat = seats[r][c]
            if seat == "L":
                if get_eight_direction_seats_count(seats, r, c) == 0:
                    new_row.append("#")
                else:
                    new_row.append("L")
            elif seat == "#":
                if get_eight_direction_seats_count(seats, r, c) >= 5:
                    new_row.append("L")
                else:
                    new_row.append("#")
            else:
                new_row.append(".")
        new_seats.append(new_row)
    return new_seats


def solve_part1(seats):
    prev_seats = seats
    new_seats = []
    occupied_count = 0
    while True:
        new_seats = get_seats_after_one_round(prev_seats)
        if prev_seats == new_seats:
            occupied_count = sum(s.count('#') for s in new_seats)
            break
        prev_seats = new_seats
    return occupied_count


def solve_part2(seats):
    prev_seats = seats
    new_seats = []
    occupied_count = 0
    while True:
        new_seats = get_seats_after_one_round2(prev_seats)
        if prev_seats == new_seats:
            occupied_count = sum(s.count('#') for s in new_seats)
            break
        prev_seats = new_seats
    return occupied_count


if __name__ == "__main__":
    print(solve_part1(seats))
    print(solve_part2(seats))

File: problem-solving/adventofcode2020/day11.cpp
// https://adventofcode.com/2020/day/11
#include <algorithm>
#include <fstream>
#include <iostream>
#include <vector>

using namespace std;

const int PART_ONE = 1;
const int PART_TWO = 2;

vector<pair<int, int> > eightDirections;

vector<vector<char> > readInput(string filepath)
{
    vector<vector<char> > graph;
    ifstream inputFile;
    inputFile.open(filepath);

    string line;
    while (getline(inputFile, line))
    {
        vector<char> row;
        for (int i = 0; i < line.length(); i++)
        {
            row.push_back(line[i]);
        }
        graph.push_back(row);
    }

    inputFile.close();
    return graph;
}

size_t getAdjacentSeatsCount(vector<vector<char> > &seats, int row, int col)
{
    size_t count = 0;
    for (pair<int, int> &d : eightDirections)
    {
        int diffRow = d.first;
        int diffCol = d.second;
        int r = row + diffRow;
        int c = col + diffCol;
        if (r >= 0 && r < seats.size() && c >= 0 && c < seats[0].size())
        {
            if (seats[r][c] == '#')
            {
                count++;
            }
        }
    }
    return count;
}

vector<vector<char> > getSeatsAfterRound(vector<vector<char> > &seats, size_t (*function)(vector<vector<char> > &, int, int), int becomeEmptyCriteria)
{
    vector<vector<char> > newSeats;
    for (int r = 0; r < seats.size(); r++)
    {
        vector<char> newRow;
        for (int c = 0; c < seats[0].size(); c++)
        {
            char seat = seats[r][c];
            if (seat == 'L')
            {
                if (function(seats, r, c) == 0)
                {
                    newRow.push_back('#');
                }
                else
                {
                    newRow.push_back('L');
                }
            }
            else if (seat == '#')
            {
                if (function(seats, r, c) >= becomeEmptyCriteria)
                {
                    newRow.push_back('L');
                }
                else
                {
                    newRow.push_back('#');
                }
            }
            else
            {
                newRow.push_back('.');
            }
        }
        newSeats.push_back(newRow);
    }
    return newSeats;
}

size_t getEightDirectionSeatsCount(vector<vector<char> > &seats, int row, int col)
{
    size_t count = 0;

    for (pair<int, int> &d : eightDirections)
    {
        int diffRow = d.first;
        int diffCol = d.second;
        int r = row + diffRow;
        int c = col + diffCol;
        while (r >= 0 && r < seats.size() && c >= 0 && c < seats[0].size())
        {
            char seat = seats[r][c];
            if (seat == '#')
            {
                count++;
                break;
            }
            else if (seat == 'L')
            {
                break;
            }
            else
            {
                r += diffRow;
                c += diffCol;
            }
        }
    }
    return count;
}

size_t solve(int part, vector<vector<char> > &seats)
{
    vector<vector<char> > prevSeats(seats);
    vector<vector<char> > newSeats;
    size_t occupiedCount = 0;
    while (true)
    {
        if (part == PART_ONE)
        {
            newSeats = getSeatsAfterRound(prevSeats, &getAdjacentSeatsCount, 4);
        }
        else if (part == PART_TWO)
        {
            newSeats = getSeatsAfterRound(prevSeats, &getEightDirectionSeatsCount, 5);
        }

        if (prevSeats == newSeats)
        {
            for (int i = 0; i < newSeats.size(); i++)
            {
                for (int j = 0; j < newSeats[0].size(); j++)
                {
                    occupiedCount += int(newSeats[i][j] == '#');
                }
            }
            break;
        }
        prevSeats = newSeats;
    }
    return occupiedCount;
}

int main()
{
    vector<vector<char> > seats = readInput("./day11.txt");

    eightDirections = {make_pair(-1, -1), make_pair(-1, 0),
                       make_pair(0, -1), make_pair(1, 1), make_pair(1, 0),
                       make_pair(0, 1), make_pair(-1, 1), make_pair(1, -1)};

    cout << solve(PART_ONE, seats) << endl;
    cout << solve(PART_TWO, seats) << endl;
}
